### PR TITLE
replacing applocal.ps1 with C++ 😍

### PR DESCRIPTION
**Pull request by valeriaconde**
Please review 🙏

### CODE REVIEW

File: azure-pipelines/e2e_projects/applocal-test/port/src/mylib.cxx
Comment (ras0219-msft) [inline azure-pipelines/e2e_projects/applocal-test/port/src/mylib.cxx:1]:
```suggestion
#define MYLIB_EXPORTS 1

#include "mylib.h"
```

File: azure-pipelines/e2e_projects/applocal-test/port/src/mylib.h
Comment (ras0219-msft) [inline azure-pipelines/e2e_projects/applocal-test/port/src/mylib.h:3]:
```suggestion
class
#if MYLIB_EXPORTS
__declspec( dllexport )
#endif
mylib {
```

**Review by ras0219-msft — COMMENTED** [review]
It would be great to avoid checking in binaries -- can you build main.exe & mylib.dll on demand instead?

File: azure-pipelines/end-to-end-tests-dir/z-applocalcpp.ps1
Comment (ras0219-msft) [inline azure-pipelines/end-to-end-tests-dir/z-applocalcpp.ps1:3]: With a separate 'build.bat' file, you should be able to do something like
```pwsh
    Run-Vcpkg env "$buildDir/build.bat" "$buildDir"
```
build.bat:
```cmd
cd %1
cl mylib.cpp /dll
cl myexe.cpp mylib.lib
```

File: src/vcpkg/commands.applocal.cpp
Comment (JackBoosY) [inline src/vcpkg/commands.applocal.cpp:128]: I'd like to make a common function instead of these, using a record list file to determine which file should be deployed.
Specific ports should not be associated with vcpkg-tool.

File: src/vcpkg/base/files.cpp
Comment (JackBoosY) [inline src/vcpkg/base/files.cpp:3016]: What about ELF(Linux) and Mach-O(OSX)?

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: src/vcpkg/base/files.cpp
Comment (JackBoosY) [reply src/vcpkg/base/files.cpp:3016]: Similar code:
ELF: https://github.com/Kitware/CMake/blob/f791cd23d408ff7af104d119898c4dd8e1e87688/Source/cmELF.cxx
Mach-O: https://github.com/Kitware/CMake/blob/f791cd23d408ff7af104d119898c4dd8e1e87688/Source/cmMachO.cxx

File: src/vcpkg/commands.applocal.cpp
Comment (valeriaconde) [reply src/vcpkg/commands.applocal.cpp:129]: I agree, we have this contemplated for the second phase of this project. This ad hoc solution is only for this first phase of my internship. I will start working on the spec for the next phase soon :)

File: src/vcpkg/base/files.cpp
Comment (dg0yt) [reply src/vcpkg/base/files.cpp:3016]: > Similar code:

And that's why I suggested a cross-platform cmake script.

Comment (BillyONeal) [reply src/vcpkg/base/files.cpp:3016]: > And that's why I suggested a cross-platform cmake script.

This needs to be usable from MSBuild where we don't have CMake available. That's part of the reason we are in this place in the first place: the current mechanism needs to defend against a zillion different powershell versions because the one vcpkg wants to supply is not known to be present.

Comment (BillyONeal) [reply src/vcpkg/base/files.cpp:3016]: > What about ELF(Linux) and Mach-O(OSX)?

`applocal.ps1` doesn't handle those so they're not in scope here.

File: src/vcpkg/base/files.cpp
Comment (BillyONeal) [reply src/vcpkg/base/files.cpp:2989]: I think the right thing is that this path should be taken and the time diff check should happen around line 3002 (after we know that last_error was `ERROR_FILE_EXISTS`)

File: src/vcpkg/base/files.cpp
Comment (BillyONeal) [inline src/vcpkg/base/files.cpp:3024]: This call isn't handing errors correctly.

File: src/vcpkg/base/files.cpp
Comment (BillyONeal) [reply src/vcpkg/base/files.cpp:3016]: Also note that CMake's implementation depends on needing to find dumpbin and is vulnerable to breaking on localized systems.

**Review by BillyONeal — CHANGES_REQUESTED** [review]

File: azure-pipelines/e2e_projects/applocal-test/port/portfile.cmake
Comment (valeriaconde) [reply azure-pipelines/e2e_projects/applocal-test/port/portfile.cmake]: I have removed this file as I am not using cmake anymore. Thanks!

File: src/vcpkg/base/files.cpp
Comment (ras0219-msft) [reply src/vcpkg/base/files.cpp:2989]: Based on my reading, the current behavior in this case would be:
1. Skips the 2989 condition
2. Checks the attributes -- if source fails, error.
3. if dest fails OR compares later than src, early return
4. CopyFileW overwrite_existing

This looks like it does the right thing to me. Are you suggesting a refactoring? Or is there a case that this doesn't handle?

File: src/vcpkg/base/files.cpp
Comment (ras0219-msft) [reply src/vcpkg/base/files.cpp:3024]: Is there a meaningful case where the file exists but we cannot obtain attributes?

Comment (BillyONeal) [reply src/vcpkg/base/files.cpp:3024]: We don't know that the destination exists here. If the call failed, `attibutes_destination` is bogus, not something that is reasonable to compare against.

Moreover, if it fails with some kind of file I/O error we need to bail and report that, not blindly ignore it.

Comment (BillyONeal) [reply src/vcpkg/base/files.cpp:3024]: That is, I believe this currently translates "failed to get destination attributes" into "OK, call `CopyFileW(... FALSE)` and destroy the destination" which is not correct.

File: src/vcpkg/base/files.cpp
Comment (autoantwort) [reply src/vcpkg/base/files.cpp]: > This needs to be usable from MSBuild where we don't have CMake available.

Just for understanding: When msbuild is able to call vcpkg to get the runtime dependencies, why is vcpkg unable to call cmake itself to get the runtime dependencies. I mean cmake must be available for vcpkg otherwise it could not build a single port.

File: src/vcpkg/base/files.cpp
Comment (valeriaconde) [reply src/vcpkg/base/files.cpp]: I have made the changes

File: src/vcpkg/base/files.cpp
Comment (valeriaconde) [reply src/vcpkg/base/files.cpp]: Thanks, I have made the changes

File: src/vcpkg/base/files.cpp
Comment (ras0219-msft) [inline src/vcpkg/base/files.cpp:3135]: My reading of these lines is that:
1. When passing `update_existing`, the file will always be truncated to 0 bytes
2. Then, timestamps will be compared. If the file is new enough, return and leave the file _empty_.

Should these two conditionals be swapped (3131 first, then 3125)? It would be great to get some tests.

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:523]: Normally, error messages need to be localized. However, since this is a z- internal command, I think it's ok to not localize this specific function.

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:47]:
```suggestion
            , m_openni2_installed(m_fs.exists(m_installed_bin_parent / "bin/OpenNI2/openni2deploy.ps1", VCPKG_LINE_INFO))
```

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:76]:
```suggestion
                Path target_item_file_path = target_binary_dir / imported_name;
```

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:117]: `Debug::print()` concatenates -- it doesn't do `printf` formatting.
```suggestion
                    Debug::println("  ", imported_name, " not found in ", m_installed_bin_parent, "; locally deployed");
```

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:118]: Should this be `installed_item_file_path` or `target_item_file_path`? It seems strange to check for the existence of one but then resolve on the other.

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
I still need to verify that each deployXYZ algorithm matches the originals, but I wanted to post this first to give you a chance to look at the other comments. Thanks!

File: src/vcpkg/commands.applocal.cpp
Comment (valeriaconde) [reply src/vcpkg/commands.applocal.cpp:117]: Thanks, I have changed all occurrences

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:120]:
```suggestion
                    Debug::println("  ", imported_name, ": ", installed_item_file_path, " not found");
```

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:413]: General suggestion: whenever you would write `const std::string&` as a function parameter, consider if `StringView` would be appropriate instead.

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:427]: This function is a bit complicated to manage the Mutant lifetime without RAII. With this suggestion, you can remove the Release/Close lines before the return statements.

```suggestion
            const auto mutant_name = "vcpkg-applocal-" + Hash::get_string_sha256(target_binary_dir);

            struct MutantGuard {
                MutantGuard(StringView name) {
                    h = ::CreateMutexW(nullptr, FALSE, Strings::to_utf16(name).c_str());
                    if (h) {
                        WaitForSingleObject(h, INFINITE);
                    }
                }
                explicit operator bool() const { return h; }
                ~MutantGuard() {
                    if (h) {
                        ReleaseMutex(h);
                        CloseHandle(h);
                    }
                }
                HANDLE h;
            };
            const MutantGuard mutant(mutant_name);

            if (!mutant)
            {
                Checks::exit_with_message("Failed to acquire mutant %s", mutant_name);
            }
```

**Review by ras0219-msft — COMMENTED** [review]
A few more suggestions, but none required so far. Still need to check the implementations against the originals. Great work!

File: src/vcpkg/commands.applocal.cpp
Comment (BillyONeal) [inline src/vcpkg/commands.applocal.cpp:419]: I don't see any reason this should be a local struct.

File: src/vcpkg/commands.applocal.cpp
Comment (BillyONeal) [inline src/vcpkg/commands.applocal.cpp:444]: IMO it would be better to just make the ctor do this which also lets you get rid of the operator bool et al.

File: src/vcpkg/base/files.cpp
Comment (BillyONeal) [inline src/vcpkg/base/files.cpp:3035]:
```suggestion
                            last_error = GetLastError();
                            ec.assign(static_cast<int>(last_error), std::system_category());
                            return false;
```

else implied by return at the end of the if block.

File: src/vcpkg/base/files.cpp
Comment (BillyONeal) [inline src/vcpkg/base/files.cpp:3128]:
```suggestion
                ec.clear();
                return false;
```

**Review by BillyONeal — APPROVED** [review]
Only nitpicks left :)

**Review by vicroms — APPROVED** [review]
LGTM. I'd like to have continuous tests for the "library specific" ones, but they worked when tested locally.

File: src/vcpkg/base/files.cpp
Comment (BillyONeal) [reply src/vcpkg/base/files.cpp]: >I mean cmake must be available for vcpkg otherwise it could not build a single port.

In classic mode (and "export" et al.) vcpkg does not build ports when called by msbuild.

File: src/vcpkg/commands.applocal.cpp
Comment (BillyONeal) [inline src/vcpkg/commands.applocal.cpp:48]:
```suggestion
```

File: src/vcpkg/commands.applocal.cpp
Comment (BillyONeal) [inline src/vcpkg/commands.applocal.cpp:57]:
```suggestion
    private:
        HANDLE h;
```

File: src/vcpkg/export.cpp
Comment (BillyONeal) [inline src/vcpkg/export.cpp:268]: @ras0219-msft is this right? (the .exe)

**Review by BillyONeal — APPROVED** [review]

File: src/vcpkg/export.cpp
Comment (valeriaconde) [reply src/vcpkg/export.cpp]: I have reverted the changes on this file as recommended by @ras0219-msft

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:290]:
```suggestion
                    m_fs.write_contents(target_binary_dir / "qt.conf", "[Paths]\n", ec);
```

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:365]:
```suggestion
                        Checks::exit_with_message(VCPKG_LINE_INFO, "qml directory must exist with Qt5Qml.dll");
```

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:367]: Is this missing a foreach? https://github.com/microsoft/vcpkg/blob/8d9c52191e039fe49c580807efee502d22f32786/ports/qt5-base/qtdeploy.ps1#L76

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:383]:
```suggestion
                for (const auto& lib : libs)
```

File: src/vcpkg/commands.applocal.cpp
Comment (ras0219-msft) [inline src/vcpkg/commands.applocal.cpp:393]: What does this line do?
1. Isn't `target_binary_name` a filename -- not a relative or absolute path to a directory? What files should this find?
2. Why is this so structurally different than the lines in qtdeploy.ps1? https://github.com/microsoft/vcpkg/blob/8d9c52191e039fe49c580807efee502d22f32786/ports/qt5-base/qtdeploy.ps1#L95

**Comment by dg0yt** [comment]
This PR hard codes informations about some plugin binaries for some ports into a binary which is maintained in another repository than the ports which provide these plugins.
- How is this meant to be kept in sync with the actual set of plugin binaries which may vary with the port versions and features?
- How is this meant to scale to the ports which provide plugins but aren't handled yet, e.g. osg/osgearth?

**Comment by vicroms** [comment]
@dg0yt 

This is Valeria's internship project; it is intended to be phase 1 of removing our dependence on PowerShell. We intend to design and implement a generic approach after merging this PR.

**Review by ras0219-msft — APPROVED** [review]
LGTM if we add deployment of both vcpkg.exe AND applocal.ps1 into export.

**Comment by ras0219-msft** [comment]
Thanks Val!

**Pull request merged**
